File: scripts/verify-single-variable-calculus-design-review.py
"""Complementary finite contract checks, not the final lesson verification suite."""

import contextlib
import io
import json
import math
from fractions import Fraction
from pathlib import Path

import mpmath as mp
from scipy.integrate import quad


data = json.loads(Path("scratch/single-variable-calculus-design-review/cases.json").read_text(encoding="utf-8"))


def close(actual, expected, tolerance=3e-12):
    assert math.isclose(actual, float(expected), rel_tol=tolerance, abs_tol=tolerance), (actual, expected)


for case in data["limitCases"]:
    epsilon, delta = Fraction(case["epsilon"]), Fraction(case["delta"])
    state = case["state"]
    jump = 2 if state["kind"] == "jump" else 0
    # Monotonic error to the right of 2 gives an open-boundary supremum.
    expected = (2 + delta) ** 2 - 4 + jump <= epsilon
    assert state["guaranteed"] == expected
    witness = state["witness"]
    if expected:
        assert witness is None
    else:
        x = Fraction(witness["inputExact"])
        displacement = Fraction(witness["displacementExact"])
        error = Fraction(witness["errorExact"])
        assert x == 2 + displacement and 0 < displacement < delta
        assert error == x * x - 4 + jump and error >= epsilon

for state in data["rates"]:
    t, h = Fraction(str(state["baseTime"])), Fraction(str(state["increment"]))
    # Expanded polynomial, rather than reusing the model's factorization.
    position = lambda x: x ** 3 - 6 * x ** 2 + 9 * x
    exact = position(t + h) - position(t)
    derivative = 3 * t * t - 12 * t + 9
    close(state["exactChange"], exact)
    close(state["secant"], exact / h)
    close(state["remainder"], exact - derivative * h)

for state in data["accumulations"]:
    upper = state["upper"]
    points = [t for t in [1, 3] if t < upper]
    signed = quad(lambda t: 3 * t * t - 12 * t + 9, 0, upper, points=points)[0]
    distance = quad(lambda t: abs(3 * t * t - 12 * t + 9), 0, upper, points=points)[0]
    close(state["exactDisplacement"], signed)
    close(state["exactDistance"], distance)
    end = Fraction(str(upper))
    width = end / state["panelCount"]
    shift = {"left": 0, "midpoint": Fraction(1, 2), "right": 1}[state["method"]]
    terms = []
    for i in range(state["panelCount"]):
        t = (i + shift) * width
        terms.append((3 * t * t - 12 * t + 9) * width)
    close(state["signedEstimate"], sum(terms))
    close(state["distanceEstimate"], sum(map(abs, terms)))

for state in data["compositions"]:
    x, h = Fraction(str(state["input"])), Fraction(str(state["increment"]))
    f = lambda x: 9 * x * x + 6 * x + 1
    close(state["exactChange"], f(x + h) - f(x))
    close(state["derivative"], 18 * x + 6)
    close(state["remainder"], f(x + h) - f(x) - (18 * x + 6) * h)

mp.mp.dps = 75
for state in data["taylors"]:
    x = mp.mpf(str(state["input"]))
    f = mp.exp if state["kind"] == "exp" else lambda x: mp.log(1 + x)
    polynomial = mp.polyval(list(reversed(mp.taylor(f, 0, state["degree"]))), x)
    close(state["polynomial"], polynomial)
    close(state["exactValue"], f(x))
    assert abs(polynomial - f(x)) <= mp.mpf(str(state["remainderBound"])) * (1 + mp.mpf("1e-12"))

programs = []
for example in data["examples"]:
    capture = io.StringIO()
    with contextlib.redirect_stdout(capture):
        exec(compile(example["code"], example["id"] + ".py", "exec"), {"__name__": "__main__"})
    actual = capture.getvalue().strip()
    assert actual == example["expected"].strip(), example["id"]
    programs.append(example["id"])

rectangle = next(e for e in data["examples"] if e["id"] == "signed-rectangles")
assert "148/9" in rectangle["expected"] and "92/9" in rectangle["expected"]
assert "coincid" in rectangle["question"]
print(json.dumps({
    "passed": True,
    "exactLimitDecisionsAndWitnesses": len(data["limitCases"]),
    "fractionSecants": len(data["rates"]),
    "quadratureAndFractionAccumulations": len(data["accumulations"]),
    "expandedCompositionChanges": len(data["compositions"]),
    "highPrecisionTaylorCases": len(data["taylors"]),
    "actualPrograms": programs,
    "signedRectangleRepairedOutput": rectangle["expected"],
    "limitations": ["No final body or lab reviewed here.", "Floating-point visual rendering of exact rational counterexamples still requires author browser checks."]
}))

File: scripts/verify-single-variable-calculus-design-review.mjs
import fs from 'node:fs';
import path from 'node:path';
import { spawnSync } from 'node:child_process';
import { createHash } from 'node:crypto';
import { limitGuarantee, motionRate, motionAccumulation, compositionChange, taylorApproximation } from '../src/learn/data/single-variable-calculus-models.js';
import { singleVariableCalculusExamples } from '../src/learn/data/single-variable-calculus-examples.js';

const directory = 'scratch/single-variable-calculus-design-review';
fs.mkdirSync(directory, { recursive: true });
const limitCases = [];
for (const kind of ['smooth', 'hole', 'jump']) {
  for (const epsilon of [.01, .03, .1, .25, .5, .83, .8399999999999999, .84, .8400000000000001, 1]) {
    for (const delta of [.001, .003, .02, .1, .19999999999999998, .2, .20000000000000004, .333, .5]) {
      limitCases.push({ epsilon: String(epsilon), delta: String(delta), state: limitGuarantee(kind, epsilon, delta) });
    }
  }
}
const rates = [];
for (const time of [1, 1.25, 1.75, 2, 2.75, 3]) {
  for (const h of [-1, -.75, -.125, -.0001, .0001, .125, .75, 1]) rates.push(motionRate(time, h));
}
const accumulations = [];
for (const upper of [.25, .75, 1, 1.5, 2.5, 3, 3.5, 4]) {
  for (const n of [1, 3, 4, 7, 8, 31]) {
    for (const rule of ['left', 'midpoint', 'right']) accumulations.push(motionAccumulation(upper, n, rule));
  }
}
const compositions = [];
for (const x of [-1, -.75, -.5, -.25, 0, .125, .75, 1]) {
  for (const h of [-.5, -.125, 0, .125, .5]) compositions.push(compositionChange(x, h));
}
const taylors = [];
for (const kind of ['exp', 'log']) {
  for (const degree of [0, 1, 2, 3, 6, 12]) {
    for (const x of [-.9, -.5, 0, .5, 1, 1.5]) taylors.push(taylorApproximation(kind, degree, x));
  }
}
fs.writeFileSync(path.join(directory, 'cases.json'), JSON.stringify({ limitCases, rates, accumulations, compositions, taylors, examples: singleVariableCalculusExamples }));
const python = process.env.LESSON_PYTHON || 'scratch/lesson-tools/Scripts/python.exe';
const result = spawnSync(python, ['-X', 'utf8', 'scripts/verify-single-variable-calculus-design-review.py'], { encoding: 'utf8' });
if (result.status !== 0) throw new Error(result.stderr || result.stdout);
const evidence = JSON.parse(result.stdout);
evidence.timestamp = new Date().toISOString();
evidence.scope = 'Initial design, pure contracts and actual displayed native programs only; not final body, visual or production verification.';
evidence.hashes = Object.fromEntries([
  'docs/teaching/SINGLE-VARIABLE-CALCULUS-LESSON-DESIGN.md',
  'src/learn/data/curriculum/blueprints/single-variable-calculus-limits-derivatives-integrals.js',
  'src/learn/data/single-variable-calculus-models.js',
  'src/learn/data/single-variable-calculus-examples.js',
].map(file => [file, createHash('sha256').update(fs.readFileSync(file)).digest('hex')]));
fs.writeFileSync(path.join(directory, 'results.json'), JSON.stringify(evidence, null, 2));
console.log(JSON.stringify(evidence, null, 2));
